xdsl/dialects/arith.py

```diff
@@ -134,16 +133,15 @@ def __hash__(self) -> int:
 @irdl_op_definition
 class Addi(BinaryOperation):
     name: str = "arith.addi"
+
     lhs: Annotated[Operand, signlessIntegerLike]
     rhs: Annotated[Operand, signlessIntegerLike]
     result: Annotated[OpResult, signlessIntegerLike]
 
-    @staticmethod
-    def get(operand1: Union[Operation, SSAValue],
-            operand2: Union[Operation, SSAValue]) -> Addi:
+    def __init__(self, operand1: Union[Operation, SSAValue],
+            operand2: Union[Operation, SSAValue]):
         operand1 = SSAValue.get(operand1)
```

Comment:
Should we do the same with `operand2`?
Also, we should probably check that the types are equal. If they are not, we should return an error rather than silently choosing one of the operands.
Otherwise, not sure if this was planned on a next PR or not, but should we just add this init in `BinaryOperation` directly?

Reply:
One might actually argue, that the verifiers are there to catch these kind of things.

Reply:
> We don't need it on operand2, since its type is not accessed wink

Oh yeah that's right, since it's modified later!

Reply:
> One might actually argue, that the verifiers are there to catch these kind of things.

I agree that verifiers are supposed to catch it later here, my problem is that it is a bit arbitrary to base the result value on the first operand.
If we are in the middle of a transformation, we would like to get notified if we are constructing an operation that has two operands with different types, and our result being the one of the first operand.
It's just a matter of raising errors now when things are arbitrary. When both operands have the same type, then it's fine returning a wrong type IMO.

Reply:
> Should we do the same with `operand2`?
> Also, we should probably check that the types are equal. If they are not, we should return an error rather than silently choosing one of the operands.
> Otherwise, not sure if this was planned on a next PR or not, but should we just add this init in `BinaryOperation` directly?

We could also check what mlir does. In this case I don't care too much, but I am a little concerned that we start to copy verifer code and basically always need to update to locations for any changes in verification.

Reply:
So it seems that MLIR does the same thing, and add the type of the first operand.
Not sure I agree with the design though, it seems quite error-prone in some corner cases.

Reply:
Also, note that I'm not in favor of adding other parts of the verifier, just the fact that I expect both operands to be the same, since we are extracting the result type based on them.
For the case where we pass the result type, we don't need to check anything.

Reply:
Why would we want to pass the result type? It is always the same as the (first) operand, no?

Reply:
It is the same, but only as long as you expect the operation to verify in the first place.
If you are in the middle of a pass, you might have operands with illegal types, that are expected to be modified later, while you want already the correct result. (This happens if you walk the module in reverse)

The way MLIR works is by defining two constructors (there are at least more of them).
The first one is the one that sets all results and operands.
The second one is only defined if the operation has a `InferResultTypes` interfaces, and will only request operands. The result types are then inferred by the interface, and then passed onto the first constructor.

To mimick this in Python, we would just have optional arguments for the results.

Reply:
OK.

Reply:
I'm okay adding an optional result_type to `__init__`, it makes sense, thanks!
Not sure, did you come to an agreement on this checking arguments type matching point?